### PR TITLE
TPP-00083: log daily symptoms

### LINKED ISSUE #83: TPP-00083: Daily Log Symptoms Card
> As a user, I want to be able to log detailed symptoms for **_the symptoms I've chosen to track_** and notes for a specific day. 
Screen Shot 2022-02-15 at 3 19 45 PM

## **Technical Information**
- Only show for symptoms the user has chosen to track so need to call GETAllTrackingPreferences() and display each accordion only if tracking=true for that symptom
- If the date already has data associated with it, the fields need to be populated to reflect this
- Tapping on "X" on top right closes the screen WITHOUT SAVING and returns to the Calendar 
- SAVE Button is always enabled, nothing is saved until SAVE button is clicked. Clicking SAVE returns to the Calendar. 
- Tapping on left and right arrows will navigate to the previous/next day 
  -  If those dates have data associated with it, the fields need to be populated as well 
- Tapping the entire section header (yellow box in image above) can open/close the accordion, and accordion's open/close status is reflected by the arrow pointing up/down 
  - More than one accordion section can be open at the same time, also possible for all accordion sections to be open, or all to be closed
  - When opening the Daily Log Symptoms Card, the Flow accordion section should be OPEN by default 
- Within each accordion section:
  - Can only make one selection
  - Tapping a selected button will de-select it
  - If one button is already selected, tapping on another button will de-select the first one and select the second one 
  - If a selection has been made, indicate that the accordion section has been filled out by changing colour of the icon and text to reflect that an option has been chosen
- Notes section should be a text box with a character limit of 500, if they try to enter more than this limit, disable the input box to prevent them from typing 
- Selections for each accordion section: 
  - Flow: NONE, LIGHT, MEDIUM, HEAVY, SPOTTING
    - Screen Shot 2022-02-19 at 5 38 30 PM

  - Mood: HAPPY, SAD, NEUTRAL, SICK, ANGRY, LOL, IDK, GREAT, LOVED
    - Screen Shot 2022-02-19 at 5 36 54 PM

  - Sleep: N/A you input a number for hours and minutes 
     - Screen Shot 2022-02-19 at 5 38 56 PM

  - Cramps: NEUTRAL, BAD, TERRIBLE, GOOD, NONE
    - Sorry the screenshot is not up to date but the buttons should look like this 
    - Screen Shot 2022-02-19 at 5 39 54 PM

  - Exercise: CARDIO, YOGA, STRENGTH, TEAM, MARTIAL ARTS, TODO (there's more) 
    - Screen Shot 2022-02-19 at 5 36 16 PM

  - Notes: N/A you input text into a textbox 
    - Ask Jing about appearance

## **Styling Requirements**
- Date should be displayed in Text between the < and > arrows 
  - clicking < or > will change that date
-

### PR DESCRIPTION
**Pull request by yuemily24**
# Description

This is a monster of a PR oop. I highly recommend you go in to test it out yourself and follow the logic as you click on things from front to back. Here's a guide as you walk through the code: 

## LogSymptomsScreen

1. Step in at `LogSymptomsScreen` (click on a random date in the calendar screen to pull it).
2. When the component mounts, it calls `GETAllTrackingPreferences` to know what accordions to display. _Note: uncomment Line 47 comment line 48 to test this because tracking prefs isn't set up yet._
     ```javascript
     const trackingPrefs = [true, true, true, true, true, true]; // TODO: remove 
     // const trackingPrefs = GETAllTrackingPreferences();
     ```
3. `useEffect` for `fetchingSymps` will kick in to fetch stored symptoms and set it to the form's symptom states. 
4. Whenever a symptom state changes, `useEffect` for `isDirty` will activate. `isDirty` knows if the form's new symtoms differs from the old symptoms. In other words, are there new symptoms to submit? 

### Selecting a symptom
Each symptom type has its own `Accordion`. `Accordion.js` is honestly a lot of just sorting through the symptoms and displaying the right components and icons for each symptom. There are 3 new components 

- `OptionPicker`: pick among some icons 

- `TextArea`: literally a text box for Notes symptom

- `TimeInput`: for exercise and sleep. Needed to import a **new package** `@react-native-picker/picker` to get a wheel selector.

Also I switched to class instead of functional components for these. If you want me to rewrite them I will...but...also :) perhaps not. 
Truthy value of symptom is the parent state, which I pass in here in `LogSymptomsScreen`. This is basically an example of nested, nested parent state passed in as props. Any changes in the children will be passed to the parent log symptoms form.
```javascript
<Accordion
     key={I}
     type={symptom}
     isLastChild={ (i === symptoms.length - 1) ? true : false }
     value={form[symptom].state} // pass in parent state
     setState={form[symptom].setState.bind(form)} // pass in parent setState function
/>
```

### Changing dates
Occurs in the arrow headers at the top. Will warn user if there is unsaved (dirty) data in the form. If user confirms they would like to switch date or there is no new data, function flow is as follows: `switchDate` -> `resetForm` -> `useEffect(() => {...}, [fetchingSymps])`.

### On submission
Pressing save button will set `submitting` state to `true`. This will trigger the submitting `useEffect`, which will POST the data to storage and close the screen. 

## Testing
Screen recording of functionalities: https://drive.google.com/file/d/1xjsBqH-tmliS7kthhc7tD2xxHiJ9RS3J/view?usp=sharing
Alert when you try to submit for a future date:
Screen Shot 2022-03-28 at 11 30 49 PM

Fixes # (issue/ticket number)

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (if required)
- [ ] My changes generate no new warnings 
**NOTE:** There are new unhandled Promise warnings but that's for Alert popup because the action is literally do nothing if they decide to close the alert. 
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by yuemily24** [comment]
NOTE TO SELF: After the frontend for tracking symptoms is merged, make sure that the call in `LogSymptomsScreen.js` doesn't bug out. Will probably have to add another state and `useEffect` tbh.
```javascript
// equivalent to componentDidMount()
useEffect(() => {
      async function fetchPreferences() {
        let trackingPrefs = await GETAllTrackingPreferences();
        // set trackingPrefs somewhere
      }
      fetchPreferences()
  }, [])
```
_EDIT: DONE!_

**Comment by faye-tan** [comment]
⚠️ Don't merge until after #92 ⚠️

**Review by kenneth-miura — APPROVED** [review]
lgtm

File: tpp-app/src/home/pages/LogSymptomsScreen.js
Comment (faye-tan) [inline tpp-app/src/home/pages/LogSymptomsScreen.js]: Correct me if wrong 😅  but GETAllTrackingPreferences uses MultiGet() https://react-native-async-storage.github.io/async-storage/docs/api/#multiget
which will resolves looking like this I believe: 
```
resolve([[TRACK_SYMPTOMS.FLOW, true],
               [TRACK_SYMPTOMS.MOOD, true],
               [TRACK_SYMPTOMS.SLEEP, true],
               [TRACK_SYMPTOMS.CRAMPS, true],
               [TRACK_SYMPTOMS.EXERCISE, true]
]); 
```
Screen Shot 2022-03-30 at 4 07 22 PM

@annie951977 this is the what the return looks like right?

File: tpp-app/src/home/pages/LogSymptomsScreen.js
Comment (faye-tan) [inline tpp-app/src/home/pages/LogSymptomsScreen.js]: See my previous comment, you will probably want to check that trackingPrefs[i][1] is true 

checking trackingPrefs[i] also relies on what's being returned from GETAllTrackingPreferences and symptoms being in the same order. For now it's fine, but I think if you matched the key and then looked at the value it would be safer.

File: tpp-app/src/home/pages/LogSymptomsScreen.js
Comment (yuemily24) [reply tpp-app/src/home/pages/LogSymptomsScreen.js]: Ah my bad, I wrote this based off the comment for `GETAllTrackingPreferences` which said it'd return an: "array of a boolean representing if a symptom is currently tracked". New commit [here](https://github.com/uoftblueprint/the-period-purse/pull/111/commits/5988d3c0badd9b5e2daa48c372fa10d49e5de1f0) reflects it now!

File: tpp-app/src/home/pages/LogSymptomsScreen.js
Comment (yuemily24) [reply tpp-app/src/home/pages/LogSymptomsScreen.js]: Changed now!

**Review by helen-li — COMMENTED** [review]
it all looks and works amazingggg!!! 🥰

just one small thing if you could double check that this also happens on your end emily -- today's april 1, and when i try to log symptoms for april 1, the error message pops up ("Sorry, this isn't a valid date") -- to my understanding we should be able to log symptoms for the current day right? 

with that said i'm not sure why this is happening tho since `isValidDate` looks like it should work just fine?

**Comment by yuemily24** [comment]
~~**DON'T MERGE YET!** Helen's right. There's some funky stuff going on in the date params. Trying to figure it out but it seems to be a strange JS Date to string conversion issue.~~

**FIXED NOW! Can be merged!**

**Review by helen-li — APPROVED** [review]
looks good!! YAY EMILY 👏

File: tpp-app/src/services/utils/helpers.js
Comment (faye-tan) [inline tpp-app/src/services/utils/helpers.js:107]: just heads up @roylin2000 pointed out this should be a synchronous function since it doesn't need to be async

Comment (faye-tan) [reply tpp-app/src/services/utils/helpers.js:107]: nvm lol i didn't see u call it anywhere

**Pull request merged**
